### DISCUSSION TITLE
30 FPS on GTX 3060

### DISCUSSION THREAD

**Discussion by ohmstead** [talmolab/sleap#1504] [Help!] [answered]
Hi there,

I am using SLEAP to track 2 nodes on 1 animal. The machine I'm using has a GTX 3060, and the data are stored locally.

My issue is that speed seems to be capped at around 30 FPS. I've tried many of @talmo's suggestions by tuning model parameters such as batch size, output stride, max stride, and filters. I'm also not tracking when I run inference.

Nothing substantially increases performance.

I know the GPU I'm using isn't comparable with the ones used in the paper, but I'm still surprised by the low inference speeds. I'm wondering if there's anything I can do to speed it up.

I tried some of the troubleshooting probes in [this thread](https://github.com/talmolab/sleap/discussions/833), and I'll include output below. One concerning thing is that with just pure frame prediction, I get about 39 FPS.

I've also attached my training_config file.

Given how simple my inference needs are (I'm trying to identify a red light & a green light in an otherwise dim frame), should I re-train a simpler model?

## troubleshooting video I/O
```
import sleap
import numpy as np
from time import perf_counter

video = sleap.load_video("video.mp4")
predictor = sleap.load_model("path/to/model_folder", batch_size=16)

# Preload images
imgs = video[:1024]

predictor.predict(imgs[:16])  # warmup

for _ in range(3):  # do multiple trials
    t0 = perf_counter()
    predictor.predict(imgs)
    elapsed = perf_counter() - t0

    fps = len(imgs) / elapsed
    print(fps)

OUTPUT
75.02424220826357
107.82811928216124
110.26847617236098
101.51743181201003
119.34829861541036
119.07180550811277
119.78881232387329
113.90664922945736
102.97892278897784
94.29292095895892
77.09119502915973
90.50859039658596
112.49337343722132
117.50251345220154
117.78927204757532
117.33310648932738
112.29441349329663
90.66583858534088
119.59791182045976
111.26626573722235
```

## troubleshooting pure prediction speed
```
video = sleap.load_video("video.mp4")
batch_size = 16
n_frames = 4096 # or len(video), but this might take a while
for i0 in range(0, n_frames, batch_size):
    t0 = perf_counter()
    i1 = min(i0 + batch_size, len(video))
    imgs = video[i0:i1]
    elapsed = perf_counter() - t0
    fps = len(imgs) / elapsed
    print(fps)

OUTPUT
2023-09-15 11:20:51.444271: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 9599 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 3060, pci bus id: 0000:01:00.0, compute capability: 8.6
Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% ETA: -:--:-- ?2023-09-15 11:21:05.7

84862: I tensorflow/stream_executor/cuda/cuda_dnn.cc:366] Loaded cuDNN version 8201
Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% ETA: 0:00:00 ?
Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% ETA: 0:00:00 41.5 FPS
38.40367086288402
Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% ETA: 0:00:00 41.8 FPS
39.10450545041766
Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% ETA: 0:00:00 41.1 FPS
39.08814592960745
```

[training_config_txt.txt](https://github.com/talmolab/sleap/files/12622971/training_config_txt.txt)

**Comment by talmo** [comment, marked as answer]
Hi @ohmstead,

The formatting on your message was a bit mangled. I tried to fix it but I'm not sure I got everything right -- I think the snippets/outputs might be switched but I left them in the order you had it.

1. As a quick sanity check: can you try `predictor.inference_model.predict` for the model benchmarking? This will help us determine the overhead from creating SLEAP objects that hold metadata.

2. For the model: I see your videos are 1440x1080, which is quite large and will likely take most of the time just transferring the data onto the GPU. You can try a couple of things to mitigate this:
    a. Decrease the input scale to 0.5 or lower, assuming you can still make out the objects you want to track in the down sampled images.
    b. Try a top down model as these will first detect the centers of the objects at lower resolution then only process a crop around them to get the full ROI.

Give these a spin and let us know if you have any questions!

Talmo

**Reply by ohmstead** [reply]
I tried both things in 2: 
a. Decreased input size to 0.5, which increased FPS by about 100-120%. Woohoo!
b. Trained models for top-down inference. Pretty similar gains in performance.

Either way, thank you for the helpful tips!

For 1, I'm still happy to try it, but not sure the exact inputs for `predictor.inference_model.predict`.

**Reply by roomrys** [reply]
Cool!

For that last bit about `InferenceModel.predict`, here is the function definition and docstring:
https://github.com/talmolab/sleap/blob/c76e6022a73ea430c7b75e7a9f5a3b79a85c5c3b/sleap/nn/inference.py#L977-L1019

which shows pretty much any data type is accepted. Since you already have the `video` loaded, you can pass in the entire `video` or just the first frame of that video with `video.get_frame(0)` as the `data`.
